### PR TITLE
MirRay OK and documented

### PR DESCRIPTION
**Pull request by EllipticEllipsis**
As Sartre didn't say, 'L'enfer c'est les virgules flottantes'.

Matched all functions, named most variables and thoroughly documented mir_ray.

Probably would have taken rather less time if I had used find and replace more carefully. Thanks to the permuter and petrie for sorting that mess out.

Very interesting actor, but has a lot of vector mathematics in it. I will document it (even) further if more detail is desired.

### CODE REVIEW

File: src/overlays/actors/ovl_Mir_Ray/z_mir_ray.h [hunk 1]
```diff
@@ -6,9 +6,44 @@
...
+    /* 0x06 */ Vec3s poolPoint; // point at center of light pool on floor for windows and BigMirror, same as source point for Cobra Mirror
+    /* 0x0C */ s16 sourceEndRadius; // Radius beam frustum at the source end
+    /* 0x0E */ s16 poolEndRadius; // Radius of beam frustum at the pool end
+    /* 0x10 */ f32 unk_10; // placement of collider center along beam
+    /* 0x14 */ s16 unk_14; // collider radius before scaled
+    /* 0x16 */ s16 lgtPtMaxRad; // light point max radius
+    /* 0x18 */ f32 unk_18; // placement of light point between source and reflection point (pool point for windows, player for mirrors)
+    /* 0x1C */ u8 r;
+    /* 0x1C */ u8 g;
+    /* 0x1C */ u8 b;
```
Comment (Roman971) [inline src/overlays/actors/ovl_Mir_Ray/z_mir_ray.h:20]: Is a color struct usable here by any chance?
```suggestion
    /* 0x1C */ Color_RGB8 color;
```

File: src/overlays/actors/ovl_Mir_Ray/z_mir_ray.c [hunk 2]
```diff
@@ -30,25 +42,537 @@ const ActorInit Mir_Ray_InitVars = {
...
+}
+
+void MirRay_Update(Actor* thisx, GlobalContext* globalCtx) {
+    s32 pad;
+    MirRay* this = THIS;
+    Player* player = PLAYER;
+
+    D_80B8E670 = 0;
+
+    if (!(this->unLit)) {
```
Comment (Roman971) [inline src/overlays/actors/ovl_Mir_Ray/z_mir_ray.c:225]:
```suggestion
    if (!this->unLit) {
```

File: src/overlays/actors/ovl_Mir_Ray/z_mir_ray.c [hunk 3]
```diff
@@ -30,25 +42,537 @@ const ActorInit Mir_Ray_InitVars = {
...
+        if (sMirRayData[this->actor.params].params & 2) {
+            if (sMirRayData[this->actor.params].params & 4) { // Beams from mirrors
+                MirRay_SetupCollider(this);
+            }
+            CollisionCheck_SetAT(globalCtx, &globalCtx->colChkCtx, &this->colliderSph.base);
+        }
+        if (0.0f < this->reflectIntensity) {
```
Comment (Roman971) [inline src/overlays/actors/ovl_Mir_Ray/z_mir_ray.c:232]:
```suggestion
        if (this->reflectIntensity > 0.0f) {
```

File: src/overlays/actors/ovl_Mir_Ray/z_mir_ray.c [hunk 4]
```diff
@@ -30,25 +42,537 @@ const ActorInit Mir_Ray_InitVars = {
...
+            CollisionCheck_SetAT(globalCtx, &globalCtx->colChkCtx, &this->shieldRay.base);
+        }
+        MirRay_MakeShieldLight(this, globalCtx);
+
+        if (0.0f < this->reflectIntensity) {
```
Comment (Roman971) [inline src/overlays/actors/ovl_Mir_Ray/z_mir_ray.c:237]:
```suggestion
        if (this->reflectIntensity > 0.0f) {
```

File: src/overlays/actors/ovl_Mir_Ray/z_mir_ray.c [hunk 5]
```diff
@@ -30,25 +42,537 @@ const ActorInit Mir_Ray_InitVars = {
...
+                    Matrix_Mult(&reflection[i].mtx, 1);
+                    gSPMatrix(POLY_XLU_DISP++, Matrix_NewMtx(globalCtx->state.gfxCtx, "../z_mir_ray.c", 1006),
+                              G_MTX_NOPUSH | G_MTX_LOAD | G_MTX_MODELVIEW);
+                    gDPSetRenderMode(POLY_XLU_DISP++, G_RM_FOG_SHADE_A, G_RM_AA_ZB_XLU_DECAL2);
+                    gDPSetPrimColor(POLY_XLU_DISP++, 0, 0, 255, 255, 150, reflection[0].opacity);
+                    gSPDisplayList(POLY_XLU_DISP++, D_060000B0);
+                }
+            }
+
+            D_80B8E670 = (u8)1U;
```
Comment (Roman971) [inline src/overlays/actors/ovl_Mir_Ray/z_mir_ray.c:511]:
```suggestion
            D_80B8E670 = 1;
```

File: src/overlays/actors/ovl_Mir_Ray/z_mir_ray.c [hunk 6]
```diff
@@ -30,25 +42,537 @@ const ActorInit Mir_Ray_InitVars = {
...
+void MirRay_Draw(Actor* thisx, GlobalContext* globalCtx) {
+    MirRay* this = THIS;
+    Player* player = PLAYER;
+    s32 i;
+    MirRayShieldReflection reflection[6];
+    s32 temp;
+
+    this->reflectIntensity = 0.0f;
+    if ((D_80B8E670 == 0) && !(this->unLit) && Player_HasMirrorShieldSetToDraw(globalCtx)) {
+        Matrix_Mult(&player->shieldMf, 0);
```
Comment (Roman971) [inline src/overlays/actors/ovl_Mir_Ray/z_mir_ray.c:473]: Any Matrix_ function that expects a `mode` argument should be called with MTXMODE_* enum values.
For instance:
```suggestion
        Matrix_Mult(&player->shieldMf, MTXMODE_NEW);
```

File: src/overlays/actors/ovl_Mir_Ray/z_mir_ray.c [hunk 7]
```diff
@@ -30,25 +42,537 @@ const ActorInit Mir_Ray_InitVars = {
...
+    closestPtz = (vecdiff.z * dist) + vecA->z;
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_Mir_Ray/MirRay_Update.s")
+    // Diameter of the double cone on the perpendicular plane through the closest point
+    coneRadius = ((radiusB - radiusA) * dist) + radiusA;
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_Mir_Ray/func_80B8D6F0.s")
+    // If the Point is within the bounding double cone, check if it is in the frustum by checking whether it is between
+    // the bounding planes
+    if (((SQ(closestPtx - pointx) + SQ(closestPty - pointy)) + SQ(closestPtz - pointz)) <= SQ(coneRadius)) {
```
Comment (Roman971) [inline src/overlays/actors/ovl_Mir_Ray/z_mir_ray.c:554]:
```suggestion
    if ((SQ(closestPtx - pointx) + SQ(closestPty - pointy) + SQ(closestPtz - pointz)) <= SQ(coneRadius)) {
```

File: src/overlays/actors/ovl_Mir_Ray/z_mir_ray.c [hunk 8]
```diff
@@ -30,25 +42,537 @@ const ActorInit Mir_Ray_InitVars = {
...
+    s32 pad;
+    MirRay* this = THIS;
+    MirRayDataEntry* temp_DataEntry = &sMirRayData[this->actor.params];
+
+    Actor_ProcessInitChain(&this->actor, sInitChain);
+    ActorShape_Init(&this->actor.shape, 0.0f, NULL, 0.0f);
+    // Generation of reflectable light!
+    osSyncPrintf("反射用 光の発生!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!\n");
+    LogUtils_LogThreadId("../z_mir_ray.c", 0x206);
+    osSyncPrintf("this->actor.arg_data = %d\n", this->actor.params);
```
Comment (Roman971) [inline src/overlays/actors/ovl_Mir_Ray/z_mir_ray.c:148]: We use LOG_* macros for these, and that last argument is a line number.
```suggestion
    LOG_NUM("this->actor.arg_data", this->actor.params, "../z_mir_ray.c", 518);
```

File: src/overlays/actors/ovl_Mir_Ray/z_mir_ray.c [hunk 9]
```diff
@@ -30,25 +42,537 @@ const ActorInit Mir_Ray_InitVars = {
...
+
+    if (this->actor.params >= 0xA) {
+        LogUtils_LogThreadId("../z_mir_ray.c", 0x209);
+        // Reflected light generation failure
+        osSyncPrintf("\"反射光 発生失敗\" = %s\n", "反射光 発生失敗");
```
Comment (Roman971) [inline src/overlays/actors/ovl_Mir_Ray/z_mir_ray.c:153]: And we also have a variant for strings
```suggestion
        // Reflected light generation failure
        LOG_STRING("反射光 発生失敗", "../z_mir_ray.c", 521);
```

**Review by Roman971 — COMMENTED** [review]
Well done with this actor, just a few things

**Review by Roman971 — APPROVED** [review]

File: src/overlays/actors/ovl_Mir_Ray/z_mir_ray.c [hunk 10]
```diff
@@ -30,25 +42,535 @@ const ActorInit Mir_Ray_InitVars = {
...
+    this->colliderSph.list->dim.worldSphere.center.y = temp_vec.y + this->sourcePt.y;
+    this->colliderSph.list->dim.worldSphere.center.z = temp_vec.z + this->sourcePt.z;
+    this->colliderSph.list->dim.worldSphere.radius = temp_DataEntry->unk_14 * this->colliderSph.list->dim.scale;
+}
+
+// Set up a light point between source point and reflection point. Reflection point is the pool point (for windows) or
+// at the player position (for mirrors) void MirRay_SetupLights(MirRay* this, GlobalContext* globalCtx) {
+void MirRay_MakeShieldLight(MirRay* this, GlobalContext* globalCtx) {
+    Player* player = PLAYER;
+    MirRayDataEntry* temp_DataEntry = &sMirRayData[this->actor.params];
```
Comment (fig02) [inline src/overlays/actors/ovl_Mir_Ray/z_mir_ray.c:108]: i would name this temp in a way that matches our style, `dataEntry` or something more speicifc if you want, but remove `temp_`

**Review by fig02 — COMMENTED** [review]
just a small thing otherwise good to go

**Review by fig02 — APPROVED** [review]

**Pull request merged**